File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    //sliding window
    int lengthOfLongestSubstring(string s) {
        int len = s.length();
        if(!len) return 0;
        int ans=0;
        int l,r;
        l=r=0;
        vector<int> freq(256,0);
        while(r<len)
        {
            freq[s[r]]++;
            while(l<=r && freq[s[r]]>1)
            {
                freq[s[l]]--;
                l++;
            }
            ans = max(ans,r-l+1);
            r++;
        }
    return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int l=0;
        int r=0;
        int[] freq = new int[256];
        int ans = 0;
        while(r<n)
        {
            freq[s.charAt(r)]++;
            while(freq[s.charAt(r)]>1)
            {
                freq[s.charAt(l)]--;
                l++;
            }
            ans = Math.max(ans,r-l+1);
            r++;
        }
        return ans;
    }
}

//time : O(n)
//space : O(1)